include/af/memory.h

```diff
@@ -99,6 +101,133 @@ AFAPI af_err af_unlock_buffer_info_event(af_event* event, af_buffer_info buf);
+
+AFAPI af_err af_memory_manager_set_initialize_fn(
+    af_memory_manager handle, af_memory_manager_initialize_fn fn);
+AFAPI af_err af_memory_manager_set_shutdown_fn(
+    af_memory_manager handle, af_memory_manager_shutdown_fn fn);
+AFAPI af_err af_memory_manager_set_alloc_fn(af_memory_manager handle,
+                                            af_memory_manager_alloc_fn fn);
+
+AFAPI af_err af_memory_manager_set_allocated_fn(
+    af_memory_manager handle, af_memory_manager_allocated_fn fn);
+AFAPI af_err af_memory_manager_set_unlock_fn(af_memory_manager handle,
+                                             af_memory_manager_unlock_fn fn);
+
+AFAPI af_err af_memory_manager_set_garbage_collect_fn(
+    af_memory_manager handle, af_memory_manager_garbage_collect_fn fn);
+AFAPI af_err af_memory_manager_set_print_info_fn(
+    af_memory_manager handle, af_memory_manager_print_info_fn fn);
+AFAPI af_err af_memory_manager_set_usage_info_fn(
+    af_memory_manager handle, af_memory_manager_usage_info_fn fn);
+
+AFAPI af_err af_memory_manager_set_user_lock_fn(
+    af_memory_manager handle, af_memory_manager_user_lock_fn fn);
+AFAPI af_err af_memory_manager_set_user_unlock_fn(
+    af_memory_manager handle, af_memory_manager_user_unlock_fn fn);
+AFAPI af_err af_memory_manager_set_is_user_locked_fn(
+    af_memory_manager handle, af_memory_manager_is_user_locked_fn fn);
+AFAPI af_err af_memory_manager_set_get_mem_step_size_fn(
+    af_memory_manager handle, af_memory_manager_get_mem_step_size_fn fn);
+AFAPI af_err af_memory_manager_set_get_max_bytes_fn(
+    af_memory_manager handle, af_memory_manager_get_max_bytes_fn fn);
+AFAPI af_err af_memory_manager_set_get_max_buffers_fn(
+    af_memory_manager handle, af_memory_manager_get_max_buffers_fn fn);
+AFAPI af_err af_memory_manager_set_set_mem_step_size_fn(
+    af_memory_manager handle, af_memory_manager_set_mem_step_size_fn fn);
+
+AFAPI af_err af_memory_manager_set_check_memory_limit(
+    af_memory_manager handle, af_memory_manager_check_memory_limit fn);
+AFAPI af_err af_memory_manager_set_add_memory_management(
+    af_memory_manager handle, af_memory_manager_add_memory_management fn);
+AFAPI af_err af_memory_manager_set_remove_memory_management(
+    af_memory_manager handle, af_memory_manager_remove_memory_management fn);
+
+////////////////////////////////////////////////////////////////////////////////
+// Native memory interface functions
+AFAPI af_err af_memory_manager_get_active_device_id(af_memory_manager handle,
+                                                    int* id);
+
+AFAPI af_err af_memory_manager_native_alloc(af_memory_manager handle,
+                                            void** ptr, size_t size);
+
+AFAPI af_err af_memory_manager_native_free(af_memory_manager handle, void* ptr);
+
+AFAPI af_err af_memory_manager_get_max_memory_size(af_memory_manager handle,
+                                                   size_t* size, int id);
```

Comment:
Not quite - these functions call functions on the `NativeMemoryInterface`, which is associated with the wrapping `MemoryManagerFunctionWrapper` because it inherits from `MemoryManagerBase`. That way, they always call the correct functions in `backend/[cpu,cuda,opencl]/memory.cpp` which properly wait on `af_events` (for the JIT) and handle the unified backend case as well without any explicit additional code.

Reply:
Will the custom memory managers have the option to not use these functions? For example, will it be okay to use cudaMalloc directly if they decide to target only the CUDA backend?

Reply:
Yes — there's no requirement that those functions be used; the internal `AllocatorInterface` doesn't actually store any state, so as long as the user is able to reproduce that internal behavior, they can do whatever they like (I've been implementing allocators that don't use these, experimentally)